task_manager/task_manager.hh

```diff
@@ -0,0 +1,362 @@
+    public:
+        task(task_impl_ptr&& impl) noexcept : _impl(std::move(impl)) {
+            register_task();
+        }
+
+        task_id id() {
+            return _impl->_status.id;
+        }
+
+        status& get_status() noexcept {
+            return _impl->_status;
+        }
+
+        uint64_t get_sequence_number() const noexcept {
+            return _impl->_status.sequence_number;
+        }
+
+        task_id get_parent_id() const noexcept {
+            return _impl->_parent_id;
+        }
+
+        void set_type(std::string type) noexcept {
+            _impl->_status.type = std::move(type);
+        }
+
+        void change_state(task_state state) noexcept {
+            _impl->_status.state = state;
+        }
+
+        void add_child(foreign_task_ptr&& child) {
+            _impl->_children.push_back(std::move(child));
+        }
+
+        void start() {
+            _impl->_status.start_time = db_clock::now();
+            _impl->_status.state = task_manager::task_state::running;
+
+            try {
+                // Background fiber does not capture task ptr, so the task can be unregistered and destroyed independently in the foreground.
+                // After the ttl expires, the task id will be used to unregister the task if that didn't happen in any other way.
+                (void)with_gate(_impl->_module->async_gate(), [f = done(), module = _impl->_module, id = id()] () mutable {
+                    return std::move(f).finally([module, id] {
+                        return sleep_abortable(module->get_task_manager().get_task_ttl(), module->abort_source());
+                    }).then_wrapped([module, id] (auto f) {
+                        f.ignore_ready_future();
+                        module->unregister_task(id);
+                    });
+                });
+                register_task();
+            } catch (...) {
+                finish_failed(std::current_exception());
+            }
+        }
+
+        module_ptr get_module() const noexcept {
+            return _impl->_module;
+        }
+
+        future<progress> get_progress() const {
+            return _impl->get_progress();
+        }
+
+        is_abortable is_abortable() const noexcept {
+            return _impl->is_abortable();
+        };
+
+        future<> abort() noexcept {
+            return _impl->abort();
+        }
+
+        future<> done() const noexcept {
+            return _impl->_done.get_shared_future();
+        }
+
+        void register_task() {
+            _impl->_module->register_task(shared_from_this());
+        }
+
+        void unregister_task() noexcept {
+            _impl->_module->unregister_task(id());
+        }
+
+        void finish() noexcept {
+            if (!_impl->_done.available()) {
+                _impl->_status.end_time = db_clock::now();
+                _impl->_status.state = task_manager::task_state::done;
+                _impl->_done.set_value();
+            }
+        }
```

Comment:
Why is this a public method? Can anybody declare the task finished?

Reply:
We usually solve this by creating a `test_task` class and make it a friend of `task`, instead of adding strange methods like this or making stuff public. See `class table` and `class column_family_test` for example.